### DISCUSSION TITLE
ParticleBinning diagnostic for pressure

### DISCUSSION THREAD

**Discussion by Tissot11** [SmileiPIC/Smilei#773] [Q&A]
I have found inconsistency while calculating `Pxx` for calculating the temperatures using the Particle Binning diagnostics. I have initialised   electrons and ions using the same temperature and density profiles. While I get the correct results for electrons `(Txx, Tyy, Tzz)` but for ions, I get incorrect results only for `Pxx or Txx`. `Tyy` and `Tzz` for ions are correct too. I have already adjusted for ion mass factor. Both the values and the profile  (`x<0`) for `Txx` component of ions is incorrect at the initialisation. `Txx` is supposed to have the same profile like for other components and other species.

![TxxEon](https://github.com/user-attachments/assets/f61622b9-0f21-4595-9286-f18fc68a1aea)
![TyyEon](https://github.com/user-attachments/assets/f48d435c-cec1-42ed-94da-63b66cd524d9)
![TzzEon](https://github.com/user-attachments/assets/99b94188-57bd-4cdf-83b6-972045411828)

![TxxIon](https://github.com/user-attachments/assets/60f2aa3e-8dcf-47c3-aeb8-1bb70c9a9150)
![TyyIon](https://github.com/user-attachments/assets/492f93a6-0dc0-4291-b905-a079303a7984)
![TzzIon](https://github.com/user-attachments/assets/6f65b0e1-1fed-4630-b2eb-f4a6eac7287e)

I am a bit perplexed by this inconsistency. How to understand this? I do have a drift in x-direction for both electron and ions. Below are the snippets for electrons and ions diagnostics

```
DiagParticleBinning(
     name = "densEon",
     deposited_quantity = "weight",
     every = everyDiag,
     time_average = 1,
     species = ["eon1", "eon2"],
     axes = [
            ["x", 0.0, Lx, 4096],
            ],
)
DiagParticleBinning(
     name = "TxxEon",
     deposited_quantity = "weight_vx_px",
     every = everyDiag,
     time_average = 1,
     species = ["eon1", "eon2"],
     axes = [
        ["x", 0.0, Lx, 4096],
           ],
)
DiagParticleBinning(
     name = "TyyEon",
     deposited_quantity = "weight_vy_py",
     every = everyDiag,
     time_average = 1,
     species = ["eon1", "eon2"],
     axes = [
          ["x", 0.0, Lx, 4096],
            ],
)
DiagParticleBinning(
     name = "TzzEon",
     deposited_quantity = "weight_vz_pz",
     every = everyDiag,
     time_average = 1,
     species = ["eon1", "eon2"],
     axes = [
          ["x", 0.0, Lx, 4096],
            ],
)
DiagParticleBinning(
     name = "densIon",
     deposited_quantity = "weight",
     every = everyDiag,
     time_average = 1,
     species = ["ion1", "ion2"],
     axes = [
          ["x", 0.0, Lx, 4096],
            ],
)
DiagParticleBinning(
     name = "TxxIon",
     deposited_quantity = "weight_vx_px",
     every = everyDiag,
     time_average = 1,
     species = ["ion1", "ion2"],
     axes = [
          ["x", 0.0, Lx, 4096],
            ],
)
DiagParticleBinning(
     name = "TyyIon",
     deposited_quantity = "weight_vy_py",
     every = everyDiag,
     time_average = 1,
     species = ["ion1", "ion2"],
     axes = [
          ["x", 0.0, Lx, 4096],
           ],
)
DiagParticleBinning(
     name = "TzzIon",
     deposited_quantity = "weight_vz_pz",
     every = everyDiag,
     time_average = 1,
     species = ["ion1", "ion2"],
     axes = [
          ["x", 0.0, Lx, 4096],
            ],
)
```

**Comment by mccoys** [comment]
Hi @Tissot11 

I moved this to the discussions as I don't think this is a bug. I believe your result is related to the fact that the definition you are using in your Pxx diagnostic (`weight_vx_px`) does not provide exactly what you seek.

`weight_vx_px` divided by `weight` essentially provides the average of $v_{x} p_{x}$ whereas the pressure is correctly defined as the average of $(v_{x}-\langle v_x\rangle) (p_{x}-\langle p_x\rangle)$.  As this latter average is equal to $\langle v_x p_x\rangle - \langle v_x\rangle\langle p_x\rangle$, one solution to your problem would be to make separate diagnostics for  $\langle v_x\rangle$ and $\langle p_x\rangle$. You can perform the operation during post-processing.

Note of course that $\langle v_x\rangle$ and $\langle p_x\rangle$ are equal to a constant factor when non-relativistic.

**Comment by Tissot11** [comment]
Thanks.  Since the default categories are either Bug or Feature Request for submitting a new issue, I choose Bug. 

Both electrons and ions have the same non-relativistic drifts (mean velocities)  in x direction, so it is a bit perplexing that electrons diagnostic gives correct results while ion diagnostics does not. Is it because, the ion momentum is scaled in terms of electrons mass scale, making it relativistic in terms of electron momentum? Of course, I can create separate diagnostics and do the stuff in post-processing.

**Reply by mccoys** [reply]
Yes the difference with electrons might be related to relativistic velocity. I am not sure though.

**Comment by Tissot11** [comment]
Do you think, this would work?
```
def stuff(particles):
    return particles.weight * (particle.vx*particles.px)
```
and subtract this `stuff` from `weight_vx_px` in post-processing. Or this `stuff` is actually `weight_vx_px` diagnostic?

**Reply by mccoys** [reply]
Your function will not work as `vx` does not exist. Note also that the `px` is a little tricky, as it does not include the mass. In fact, this `px` is only the lorentz factor multiplied by the velocity.

Anyways, what you have here is the same as `weight_vx_px`. Instead, you need another diagnostic that simply uses `weight_px`.

**Comment by Tissot11** [comment]
Yeah, I realised it after trying this. I dump a 2D diagnostic `xPx` which have both `px` and `x` and I save `weight.` Doing the same for `xVx` so `x` and `vx` and saving `weight` should work as the second part in your expression above?

**Reply by mccoys** [reply]
I don't understand. You want `x` and `px` as axes? This is not what you should do. You must have all diagnostics with the same axes (`x` in your case). Then one diagnostic has `deposited_quantity=weight_vx_px` and another one has `deposited_quantity=weight_px`

**Comment by Tissot11** [comment]
This will help only getting `<px>`, but how to get `<vx>` then? I need then another diagnostic for `gamma` vs `x`? To be precise, could you please tell how to calculate the second term in the expression you wrote above and whether I should use it a a general procedure for all species?

**Reply by mccoys** [reply]
If this is non relativistic, px and vx are the same. If you are considering a species with fixed charge, you can actually use `weight_charge_vx` to get vx. Otherwise you need a custom function like `lambda p: p.weight * p.px / (1+p.px**2+p.py**2+p.pz**2)**0.5`

**Reply by mccoys** [reply]
Note that using a custom function prevents changing units in happi

**Comment by Tissot11** [comment]
I designed two more diagnostics (`numDiags=256`)


```
DiagParticleBinning(
     name = "densIon",
     deposited_quantity = "weight",
     every = everyDiag,
     time_average = 1,
     species = ["ion"],
     axes = [
          ["x", 0.0, Lx, numDiags],
            ],
)
DiagParticleBinning(
     name = "TxxIon",
     deposited_quantity = "weight_vx_px",
     every = everyDiag,
     time_average = 1,
     species = ["ion"],
     axes = [
          ["x", 0.0, Lx, numDiags],
            ],
)

DiagParticleBinning(
     name = "pxIon",
     deposited_quantity = "weight_px",
     every = everyDiag,
     time_average = 1,
     species = ["ion"],
     axes = [
          ["x", 0.0, Lx, numDiags],
            ],
)
DiagParticleBinning(
     name = "vxIon",
     deposited_quantity = "weight_charge_vx",
     every = everyDiag,
     time_average = 1,
     species = ["ion"],
     axes = [
          ["x", 0.0, Lx, numDiags],
            ],
)```


and use this 


```pressure = S.ParticleBinning("TxxIon", timesteps=timeSteps[timesteps])
        pressureData = np.array(pressure.getData()[0]).T
        
        pxDiag = S.ParticleBinning("pxIon", timesteps=timeSteps[timesteps])
        pxData = np.array(pxDiag.getData()[0]).T / massratio
        
        vxDiag = S.ParticleBinning("vxIon", timesteps=timeSteps[timesteps])
        vxData = np.array(vxDiag.getData()[0]).T
        
        temperature = (pressureData - pxData* vxData) / (massRatio*densData)
```

It again produces


![TxxIonCorr](https://github.com/user-attachments/assets/90469d75-208e-4d63-a7ee-89f937a2f613)


it is the same as before.

if I use 

```pressure = S.ParticleBinning("TxxIon", timesteps=timeSteps[timesteps])
        pressureData = np.array(pressure.getData()[0]).T
        
        pxDiag = S.ParticleBinning("pxIon", timesteps=timeSteps[timesteps])
        pxData = np.array(pxDiag.getData()[0]).T 
        
        vxDiag = S.ParticleBinning("vxIon", timesteps=timeSteps[timesteps])
        vxData = np.array(vxDiag.getData()[0]).T
        
        temperature = (pressureData - pxData* vxData) / (massRatio*densData)
```

I get even a funny result



![TxxIonCorr1](https://github.com/user-attachments/assets/aa326931-4f22-4b44-a623-344157e3c504)



 So something is really strange here for ion species with a non-relativistic drift in `x` direction. I request you to kindly look into it carefully and I hope it is not a bug..


Please note that I made slight changes in the `Namelist`, so the fluctuations on the right side are a bit smaller than before.

**Reply by mccoys** [reply]
The mass must divide the diags with px. You forgot to divide the px and vx dishes individually by the weight diag. And you also forgot to divide one by the charge

**Comment by Tissot11** [comment]
I am sorry but I am not following you here. Charge is +1, so there is not need to worry about. But what do you mean divide by `weight`?

Could you please take my expression and write it clearly...

**Reply by mccoys** [reply]
temperature = (pressureData/densdata - pxData/densdata* vxData/densdata) / massRatio

**Comment by Tissot11** [comment]
Since I initialize temperature in the units of `electron rest mass energy,` so I do not need to divide it by `mass ratio` for checking the results. After realising `weight_charge_vx ` is indeed current density in `x`, I rescaled it (as you suggested) as `temperature = ( pressureData  - pxData* (vxData/densData)) / (densData)`. This indeed gives the correct result.
